### DISCUSSION TITLE
Struggling with connecting multiple 3.5 RPi displays

### DISCUSSION THREAD

**Discussion by cvels** [Bodmer/TFT_eSPI#1947] [Q&A - General]
Hi Bodmer,

I’m struggling with connecting multiple 3.5 displays to a NodeMCU ESP32 dev board.
I was only able to get it working when using 2 TFT SPI displays, based on the ILI9431 driver. (https://www.waveshare.com/wiki/2.4inch_LCD_Module )
The problem is that the ILI9431 (at least that is what I found on the net) is limited to a display size of something like 2.8inch.
My second attempt is to have 2 TFT SPI 3.5 inch displays also from Waveshare (https://www.waveshare.com/3.5inch-tft-touch-shield.htm ) working together on the same SPI bus. These displays are using driver chipset ILI9486 (RPI_ILI9486).
I’m using the CS trick to distinguish between the displays in the sketch, but for some reasons seems that the display is not picking up this.
At the moment, I’m only trying to have one display working …
When I define CS correctly in the setup11_RPi_touch_ILI9486.h, it’s all working. (I have tried different ways with user_setup.h where I defined RPI_ILI0486_Driver etc.
But when I define CS as -1 and do the CS handling in the sketch, the display stays white and not reacting at all.
In your code, I saw remarks about an RPi display is reacting slower/different on CS changes … maybe related.

IDE: Arduino
TFT_eSPI version: 2.4.72
interface SPI:

Any idea on  what I’m doing wrong?
Many thx in advance,
Kind regards,
Carl
[test TFT.txt](https://github.com/Bodmer/TFT_eSPI/files/9242402/test.TFT.txt)
[Setup11_RPi_touch_ILI9486.txt](https://github.com/Bodmer/TFT_eSPI/files/9242414/Setup11_RPi_touch_ILI9486.txt)
l

**Comment by Bodmer** [comment]
I don't think this is an issue with the library is the display operates when CS is defined.

I would expect the display to operate if the CS is controlled by the sketch. Note that the chip select must be low during tft.init() or tft.begin().

**Reply by cvels** [reply]
yes, i expected the same ... but seems that for some reason the display is not reacting ... 
or i'm I doing something wrong? you have more experience in this than me.
What can i do to dive deeper ?

#define SCH2_TFT_CS 15
#define SCH2_CS_L GPIO.out_w1ts = (1 << SCH2_TFT_CS); GPIO.out_w1tc = (1 << SCH2_TFT_CS);delay(100)
#define SCH2_CS_H GPIO.out_w1tc = (1 << SCH2_TFT_CS); GPIO.out_w1ts = (1 << SCH2_TFT_CS);delay(100)

SCH2_CS_L;
tft.init();
tft.setTextSize(5);
SCH2_CS_H;

SCH2_CS_L;
tft.fillScreen(TFT_YELLOW);
tft.print("Hello world1");
digitalWrite(scherm2, HIGH);
SCH2_CS_H;

is the above correct ?

**Reply by Frank-Bemelman** [reply]
The way digitalWrite does is :GPIO.out_w1ts = ((uint32_t)1 << pin);

But I prefer something easier to read. It's not that you need the speed, with that delay(100) you added.
I would write a function something like this (not tested):

```
void SelectDisplay( uint8_t display)
{ // deselect all displays 
  digitalWrite(15, HIGH);
  digitalWrite(16, HIGH); // whatever pins you use for the other displays
  digitalWrite(17, HIGH); // whatever pins you use for the other displays

  switch(display){
    case 0: 
      digitalWrite(15, LOW);
      break;
  case 1:
      digitalWrite(16, LOW); // whatever pin you use for CS
      break;
  case 2:
      digitalWrite(17, LOW); // whatever pin you use for CS
      break;
 default:
      digitalWrite(15, LOW);
      break;
  }
} 
```

Other than that, you could write a test program that prints to each screen, changing every 5 seconds, And then check with a voltmeter what your CS lines are at. It should be a solid logic low close to 0V or logic high close to 3V3.

``

**Reply by cvels** [reply]
Thx for the feedback Frank. I will try it immediately when at home tomorrow. Keep you posted.

**Reply by cvels** [reply]
Using kind of SelectDisplay function is resulting in the same (a bit as expected). It's for sure usefull idea as soon i'm able to manage the CS line by meself resulting in working TFT's.

results of the 'test program', toggling every 5 seconds between TFT1 and TFT2:
- Measuring output CS pin, LOW = -0,04099 V + HIGH = 3,2499 V
                     VIN -> 5V of Display = 4,58V

Does anyony knows if these levels are enough ? (but they should be the same when managing CS via the library)

weird ?

**Reply by Frank-Bemelman** [reply]
Looks allright to me. Just start with one display, once that works, add another. 
Nothing weird.

**Reply by cvels** [reply]
That's exactly what i'm doing. Only one display. Cabling is ok, because when using the #define CS in the library it's working. But as soon i change the #define CS to -1 and taking care in the sketch about the CS level ... it's not working anymore, while  it should.

next steps for me are:
1) change the CS_L and CS_H in the library (as test) to a TFT_eSPI member function which is controlling the CS level. That should even more close on what the library is doing standard.
2) cost a bit more time, as i'm not that used to it. Is using my old logic analyser to understand what the differences are.

I'll keep you posted.

**Reply by Frank-Bemelman** [reply]
If you have working setup with one display (CS defined to some pin) - I would expect the display would also work if you connect CS to ground, not to the pin. And then you could try again, connecting CS to a pin of your choice, and control it from the sketch. I would expect that all should work. Last step would then be to define CS as -1 and see if it keeps working. And add a second display.. That would be my approach to the problem you seem to have.

**Reply by cvels** [reply]
Hi Frank, I agree but that is the weird thing. I tried that before creating this topic, connecting the CS pin directly to the ground was also not helping. I'm missing something small and stupid ... but cannot find the light. That's why i have ask the question in the first place to find out if the library was doing somewhere a magic thing which makes the difference. And it's only with this display/driver (the other displays with ILI9431 are working as a charm)

**Reply by Bodmer** [reply]
I suspect the problem may be related to the serial to parallel interface on these RPi boards. The chip select manages a counter that generates the write pulse.

To check if the power demand and SPI bus loading is not the issue I would connect both displays to the exact same pins (including the same chip select pin) and check both displays show the same thing.

**Reply by cvels** [reply]
Hi Bodmer, I have connected both displays together using exact same pins (also CS).  The result is the same.
When using your library, both screen shows the same content (= ok). When handling the CS inside the sketch (and put in the header files of the library "#define TFT_CS   -1" ) both screens turns white. No content.

**Reply by Bodmer** [reply]
This shows that the RPi interface chip select cannot be controlled from the sketch. This is likely due to the way the serial to parallel interface in these designs generate the control signals to the display.  It appears that there is no work-around for this behaviour.
